OSDOCS-12321# GCP Filestore WIF support (#87736)
openshift/openshift-docs: modules/persistent-storage-csi-gcp-filestore-wif.adoc

// Module included in the following assemblies:
//
// * storage/container_storage_csi-google_cloud_file.adoc

:_mod-docs-content-type: PROCEDURE
[id="persistent-storage-csi-gcp-filestore-wif_{context}"]
= Preparing to install the {gcp-short} Filestore CSI Driver Operator with Workload Identity

> If you are planning to use {gcp-wid-short} with Google Compute Platform Filestore, you must obtain certain parameters that you will use during the installation of the {gcp-short} Filestore Container Storage Interface (CSI) Driver Operator.

Reviewer: ```suggestion
If you are planning to use {gcp-wid-short} with {gcp-full} Filestore, you must obtain certain parameters that you will use during the installation of the {gcp-short} Filestore Container Storage Interface (CSI) Driver Operator.
```

.Prerequisites
* Access to the cluster as a user with the cluster-admin role.

// Put note in install area of docs to remind users to take note of the identity pool ID and the provider ID

.Procedure

> To prepare to install the {gcp-short} Filestore CSI Driver Operator with Workload Identity:

Reviewer: Shouldn't have plain text in the steps area (and also we should not use a trailing sentence as a list intro per accessibility guidelines). You could move this to the intro and finish the sentence, but I think it'a already covered by the intro you have :slightly_smiling_face: 
```suggestion
```

. Obtain the project number:

.. Obtain the project ID by running the following command:
+
[source, terminal]
----
$ export PROJECT_ID=$(oc get infrastructure/cluster -o jsonpath='{.status.platformStatus.gcp.projectID}')
----

.. Obtain the project number, using the project ID, by running the following command:
+
[source, terminal]
----
$ gcloud projects describe $PROJECT_ID --format="value(projectNumber)"
----

. Find the identity pool ID and the provider ID:
+

> During cluster installation, the names of these resources is provided to `ccoctl` with the `--name parameter`. See Step 3 of Section _Creating {gcp-short} resources with the Cloud Credential Operator utility_.

Reviewer: ```suggestion
During cluster installation, the names of these resources are provided to the Cloud Credential Operator utility (`ccoctl`) with the `--name parameter`. See "Creating {gcp-short} resources with the Cloud Credential Operator utility".
```
(would avoid using the actual number as it can change later and make this inaccurate)

Author: Fixed.

. Create Workload Identity resources for the {gcp-short} Filestore Operator:

.. Create a `CredentialsRequest` file using the following example file:
+
.Example Credentials Request YAML file
[source, YAML]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: openshift-gcp-filestore-csi-driver-operator
  namespace: openshift-cloud-credential-operator
  annotations:
	include.release.openshift.io/self-managed-high-availability: "true"
	include.release.openshift.io/single-node-developer: "true"
spec:
  serviceAccountNames:
  - gcp-filestore-csi-driver-operator
  - gcp-filestore-csi-driver-controller-sa
  secretRef:
	name: gcp-filestore-cloud-credentials

> 	namespace: openshift-cluster-csi-drivers

Reviewer 2: It seems that there is a Tab space.

Author: Fixed. Thanks!

  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
	kind: GCPProviderSpec
    predefinedRoles:
    - roles/file.editor

>     - roles/resourcemanager.tagUser

Reviewer 2: Maybe we also need this one `roles/iam.workloadIdentityUser` , let me research it more.

Author: @chao007 Thanks! Please let me know.

Reviewer 3: I believe ccoctl will manage roles/iam.workloadIdentityUser in the next step.

Reviewer 2: Yes I think so.

Author: @jstuever No changes needed then?

Reviewer 3: Correct.

Author: Confirmed by @jstuever. Closing comment.

	skipServiceCheck: true
----

.. Use the `CredentialsRequest` file to create a {gcp-short} service account by running the following command:
+
[source, terminal]
----

> $ ./ccoctl gcp create-service-accounts --name=<filestore-service-account> <1>
> --workload-identity-pool=<workload-identity-pool> <2> 
> --workload-identity-provider=<workload-identity-provider> <3> 
> --project=<project-id> <4> 
> --credentials-requests-dir=/tmp/credreq <5>

Reviewer: format issues here should be resolved - this will paste incorrectly as-is
```suggestion
$ ./ccoctl gcp create-service-accounts --name=<filestore-service-account> \// <1>
  --workload-identity-pool=<workload-identity-pool> \// <2> 
  --workload-identity-provider=<workload-identity-provider> \// <3> 
  --project=<project-id> \// <4> 
  --credentials-requests-dir=/tmp/credreq <5>
```

Author: Fixed.

Reviewer: This is now indented, but doesn't have the required format fix to make this paste correctly. I'll leave a new comment with the fix

> $ ./ccoctl gcp create-service-accounts --name=<filestore-service-account> <1>
>   --workload-identity-pool=<workload-identity-pool> <2> 
>   --workload-identity-provider=<workload-identity-provider> <3> 
>   --project=<project-id> <4> 
>   --credentials-requests-dir=/tmp/credreq <5>

Reviewer: This is required or it will not copy/paste right
```suggestion
$ ./ccoctl gcp create-service-accounts --name=<filestore-service-account> \// <1>
  --workload-identity-pool=<workload-identity-pool> \// <2> 
  --workload-identity-provider=<workload-identity-provider> \// <3> 
  --project=<project-id> \// <4> 
  --credentials-requests-dir=/tmp/credreq <5>
```

Author: Fixed.

----

> <1> <filestore-service-account> is a user-chosen name. 
> <2> <workload-identity-pool> comes from Step 2 above.
> <3> <workload-identity-provider> comes from Step 2 above.
> <4> <project-id> comes from Step 1.a above.

Reviewer: You've just had them get these values, and called them by name in those steps. Can refer to the actual items and not reference steps that might change to avoid doc maintenance toil. 
```suggestion
<1> `<filestore-service-account>` is a user-chosen name. 
<2> `<workload-identity-pool>` is the identity pool ID.
<3> `<workload-identity-provider>` is the provider ID.
<4> `<project-id>` is the project ID.
```

> <4> <project-id> comes from Step 1.a above.

Reviewer: All user-replacable values should use underscores between words, not hyphens. e.g. `<filestore_service_account>`

<5> The name of directory where the `CredentialsRequest` file resides.
+
.Example output
[source, terminal]
----
2025/02/10 17:47:39 Credentials loaded from gcloud CLI defaults
2025/02/10 17:47:42 IAM service account filestore-service-account-openshift-gcp-filestore-csi-driver-operator created
2025/02/10 17:47:44 Unable to add predefined roles to IAM service account, retrying...
2025/02/10 17:47:59 Updated policy bindings for IAM service account filestore-service-account-openshift-gcp-filestore-csi-driver-operator
2025/02/10 17:47:59 Saved credentials configuration to: /tmp/install-dir/ <1>
openshift-cluster-csi-drivers-gcp-filestore-cloud-credentials-credentials.yaml
----
<1> The current directory.

.. Find the service account email of the newly created service account by running the following command:
+
[source, terminal]
----

> cat /tmp/install-dir/manifests/openshift-cluster-csi-drivers-gcp-filestore-cloud-credentials-credentials.yaml | yq '.data["service_account.json"]' | base64 -d | jq '.service_account_impersonation_url'

Reviewer: ```suggestion
$ cat /tmp/install-dir/manifests/openshift-cluster-csi-drivers-gcp-filestore-cloud-credentials-credentials.yaml | yq '.data["service_account.json"]' | base64 -d | jq '.service_account_impersonation_url'
```

Author: Fixed.

----
+
.Example output
[source, terminal]
----
https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/filestore-se-openshift-g-ch8cm@openshift-gce-devel.iam.gserviceaccount.com:generateAccessToken
----
+

> In this example output, the service account email is: filestore-se-openshift-g-ch8cm@openshift-gce-devel.iam.gserviceaccount.com.

Reviewer: ```suggestion
In this example output, the service account email is `filestore-se-openshift-g-ch8cm@openshift-gce-devel.iam.gserviceaccount.com`.
```

Author: Fixed.

.Results

You now have the following parameters that you need to install the {gcp-short} Filestore CSI Driver Operator:

* Project number - from Step 1.b

* Pool ID - from Step 2

* Provider ID - from Step 2

> * Service account email - from Step 3.c

Reviewer: Would remove this section, as it's pretty straightforward. If it's really needed, I would at least remove the specific step reference to avoid extra docs maintenance work
